### DISCUSSION TITLE
Maintaining 'Creating Your Project in Drasil' Wiki Tutorial

### DISCUSSION THREAD

**Discussion by peter-michalski** [JacquesCarette/Drasil#3194] [General]
This discussion is meant to build off of #3072, where I updated the [wiki tutorial](https://github.com/JacquesCarette/Drasil/wiki/Creating-Your-Project-in-Drasil) to reflect the current code base. 

Prior to the changes that were implemented in that issue, the tutorial page had not been updated in well over a year's time. 

Some of the items that were addressed included

- out of date paths
- missing imports
- missing numbered list items
- lack of links to Haddock documentation; these could help clarify any confusion regarding sections and constructors
- insufficient information about ChunkDB
- insufficient information about symbMap

The addition of links throughout  the tutorial to relevant sections of the Haddock documentation should really help with understandability. 

I suggest the following moving forward:

1. While I don't think it's practical to have developers check this tutorial every time they commit code changes, it would be nice to have them or the PR reviewer be mindful of and consider if the code changes impact how examples are built and then update /flag the tutorial if appropriate.
2. Perhaps an annual or semi-annual building of an example, by a designated Drasil team member, using the steps of the tutorial. We can then catch any issues in the tutorial that might have arisen over the course of the year.
3. New members on the project should build a toy example using the tutorial and provide feedback.
4. Any CAS 741 students that choose to build a Drasil example could be asked to provide feedback about the tutorial.

**Comment by smiths** [comment]
Great comments @peter-michalski.  I agree.  I currently have 17 students registered in CAS741.  Although the class will likely get smaller, I'm optimistic that there will still be some that wish to try out Drasil.  The last time I taught CAS 741 there was at least 3 students (including @tingyuw) that chose to use Drasil for their project.  It will be easier to attract students to try Drasil now that we can solve higher dimension ODEs.  This will be a good test for our wiki tutorial.

**Comment by JacquesCarette** [comment]
Hmm, on top of using periodic human-driven QA passes to keep the tutorial up to date, we could also put comments in the appropriate places in the Drasil source code that would direct Drasil developers to double-check the tutorial if they change certain parts.

Another (additional!) possibility would be to make the tutorial not part of the wiki but a generated (by Drasil) set of web pages. Then we could do some automated testing of it.

I don't see my comments as being XOR, but rather as being 'AND-ALSO' comments.

Lastly: thanks @peter-michalski for being quite specific about what was broken in the tutorial. That helps focus the solution space.

**Reply by smiths** [reply]
Great ideas @JacquesCarette.  @peter-michalski, maybe you could suggest some specific spots in the Drasil source code for the comments to the developers?  If we don't act on the idea from @JacquesCarette, the idea could get lost over time.

The [web-page that lists of Drasil case studies](https://jacquescarette.github.io/Drasil/) is currently [generated by Drasil](https://github.com/JacquesCarette/Drasil/tree/master/code/drasil-website), so the idea of generating the tutorial as a set of webpages is feasible.

**Reply by peter-michalski** [reply]
@smiths The following spots in the Drasil source code come to mind:

- /code/drasil-example/template/lib/Drasil/Template/Body.hs 
- /code/drasil-database/lib/Database/Drasil/ChunkDB.hs
- /code/drasil-sysinfo/lib/SysInfo/Drasil/SystemInformation.hs
- /code/drasil-docLang/lib/Drasil/DocLang/SRS.hs
- /code/drasil-docLang/lib/Drasil/DocDecl.hs
- /code/drasil-data/lib/Data/Drasil/Concepts/Documentation.hs

While changes to many files can result in the tutorial going out of date, the above files include crucial information for building examples and any changes to them should consider affects on the tutorial.

**Reply by smiths** [reply]
@peter-michalski, thank you for following up on this.  Maybe we can add a comment at the top of each file instructing anyone making changes to the file to check if the changes require editing the tutorial?

**Reply by peter-michalski** [reply]
@smiths, I added comments in #3205

**Reply by balacij** [reply]
> Another (additional!) possibility would be to make the tutorial not part of the wiki but a generated (by Drasil) set of web pages. Then we could do some automated testing of it.

We don't necessarily need to go for the full "generated by Drasil" yet too, we would really just add it somehow to the existing website, and then the PR process would catch instances of "Drasil code" changes in a PR that may affect Drasil users.

**Reply by smiths** [reply]
I like the idea of the PR process catching changes that may require special attention.

Another idea that comes to mind is to use a [customized issue template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/configuring-issue-templates-for-your-repository).  The person creating the issue could then be asked if the changes will affect the tutorial wiki.  I'm not actually in favour of this option, since I think issue templates are too heavy handed for our work flow.  :smile:

**Reply by JacquesCarette** [reply]
I agree, the templates push things too far. Having the PR process catch more is definitely a good idea.

**Comment by balacij** [comment]
So, another idea came to mind today in CAS 741 (similar line of thinking to the [above thread](https://github.com/JacquesCarette/Drasil/discussions/3194#discussioncomment-4478465)): we can strip down the current document to something barebones, keeping it very generic, only explaining the general workflow, and then build out the `template` project a bit more to be a somewhat interactive guide to building a simple case study that generates code+srs. The `template` is currently just a "skeleton", but we can probably use it as a guided example of how to build a really small example. The example could be something with a very low amount of information, but be built using a textual commentary. 

Doing it this way would ensure that the code found in the example would be more up-to-date because we test the template Haskell code compilation (I think the SRS generation too?) as part of the CI we use to review PRs. Thus, we'll be forced to ensure that the template has working code. If the template code breaks, then we'll also be forced to look at the comments too.

Another bonus of this is that we don't need to have the Haskell comments in our codebase noting to update areas of the Wiki, and users also have their IDE tooling immediately available, allowing them to get a better idea of other examples and such faster. Overall, I think it would decrease the learning curve a bit and perhaps make learning faster?

In some sense, it is adding a bit of "trial and error" learning, which most devs do at some point anyways.

The current "Creating your project in Drasil" document would become, basically, a tutorial on copying and pasting the `template` project, renaming the resultant folder into their project name, and then editing it, learning along the way. The majority of the "building" and "testing" is instrumented by the Makefile anways, which is already documented elsewhere.

**Reply by balacij** [reply]
I suppose this would be learning about Drasil in a manner akin to Jupyter notebooks teaching about problems?

**Reply by JacquesCarette** [reply]
I think there's room for a template which is just a skeleton, **and** a guided interactive guide. They serve different purposes.

**Reply by balacij** [reply]
Yep, that makes sense. Experts would probably prefer to not waste time deleting the guide portion of a "guided interactive template." In the future, it might be nice to have a `drasil init`-like project starter (and hence a `drasil` program).

Another one of the benefits of the guided interactive template is that it gets the ball rolling faster. Users can quickly see how changes propagate around in Drasil, how their changes affect SRS+code generation, etc. We can force them to see and learn what Drasil is about + learn about the current workflow at the same time.

The guided template would probably need to have a shallow amount of information, a minimum viable amount to get a user going, and then directing them to the other case studies or the Haddock docs for extra information and such. They'll have their IDE too, so I would hope that they feel comfortable poking around the code too, to learn about things on their own.
